### DISCUSSION TITLE
Best practices for test naming requirements

### DISCUSSION THREAD

**Discussion by dotnwat** [mesonbuild/meson#11275] [General]
We have hundreds of tests, and have ended up with this pattern:

```
t = exectuable('exe_name', 'source.cc', ...)
test('test name', t)
```

However, this pattern requires me to choose 3 different names (variable, test name, exe name). In practice I really only think it should be necessary to name the executable. For the variable I just re-use the same name `t` everhwere, which I assume is ok, but feels awkward.

I'm curious if there is a better pattern to use here?

Thanks!

**Comment by eli-schwartz** [comment]
One common approach assumes every executable has one source named the same as the test, and is run without arguments:

```meson
tests = [
    'foo',
    'bar',
    'baz',
]

foreach t: tests
    exe = executable(t, t+'.c', dependencies: [commondep])
    test(t, exe)
endforeach
```

**Reply by dotnwat** [reply]
thanks @eli-schwartz yeh we do something similar for managing the large number of tests today. but wasn't sure there was something i was missing out on.

**Reply by eli-schwartz** [reply]
I guess maybe `test_exes += {t, exe}` can be used in addition if you anticipate needing to do something more complex than that on occasion. You can also make `tests` into an array of arrays, and have the second element be `{}` or possibly contain optional arguments to reuse inside the foreach.

**Comment by bruchar1** [comment]
I don't see anything wrong with that. You may use the same name for the executable and the test if it fits for you. But there are cases where you may want to create different test targets from the same executable (passing different arguments to the executable for instance), and this is one reason why a distinct test name can be useful.

**Reply by dotnwat** [reply]
> want to create different test targets from the same executable (passing different arguments to the executable for instance), and this is one reason why a distinct test name can be useful.

excellent point @bruchar1 this didn't occur to me. thanks!

**Reply by eli-schwartz** [reply]
You can stick each executable into a:
```meson
test_exes = {}
test_exes += {t, exe}
```

And then you can look them up by name in the dictionary (and also check if a given name is present).


(Alternatively, it's possible to use set_variable to programmatically determine the variable token to use.)
